### PR TITLE
Create testing

### PR DESCRIPTION
**Pull request by quothbonney**
Created a basic testing structure. Note the new directory `dependencies/` which links some relevant deps for testing. The main change is the new directory `testing/` which contains a `CMakeLists.txt` as well as individual tests.

By far the most complex test was ensuring that the PolySelect will never fail and never create intersecting polygons.

### CODE REVIEW

**Review by quothbonney — COMMENTED** [review]
Need to get merged ASAP

File: tests/raster_test.cpp [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+                int randY = std::rand() % maxSize;
+                vec.push_back(point{randX, randY});
+            }
+            CHECK(intersections(vec) == 1);
+        }
+    }
+
+
+
+}
```
Comment (c-bik) [inline tests/raster_test.cpp:92]: Remove empty lines
Add a new line at EOF

File: tests/main.cpp [hunk 2]
```diff
@@ -0,0 +1,2 @@
+#define CATCH_CONFIG_MAIN
+#include <catch2/catch.hpp>
```
Comment (c-bik) [inline tests/main.cpp:2]: New line at EOF

File: tests/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+project(tests)
+
+
+add_executable(tests
+        main.cpp
+        raster_test.cpp
+        ../src/Raster.cpp
+        ../src/PolySelect.cpp)
+target_link_libraries(tests PRIVATE Catch2::Catch2 GDAL::GDAL ${Eigen_LIBRARIES})
+
```
Comment (c-bik) [inline tests/CMakeLists.txt:11]: Remove empty line

File: tests/raster_test.cpp [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+            std::cout << p2.x << " " << p2.y << "   " << q2.x << " " << q2.y << "\n\n\n";*/
+        }
+
+    }
+    return false;
+}
+
+TEST_CASE("Raster Test") {
+    const char *ok = "../../data/oklahoma.tif";
+    const char *ga = "../../data/gabon.tif";
```
Comment (c-bik) [inline tests/raster_test.cpp:62]: Use parameters for test files instead of hardcoding.

File: tests/raster_test.cpp [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+            // Special Cases
+            if (o1 == 0 && onSegment(p1, p2, q1)) return true;
+            if (o2 == 0 && onSegment(p1, q2, q1)) return true;
+            if (o3 == 0 && onSegment(p2, p1, q2)) return true;
+            if (o4 == 0 && onSegment(p2, q1, q2)) return true;
+            /*
+            print testing;
+            std::cout << p1.x << " " << p1.y << "   " << q1.x << " " << q1.y << "\n";
+            std::cout << p2.x << " " << p2.y << "   " << q2.x << " " << q2.y << "\n\n\n";*/
```
Comment (c-bik) [inline tests/raster_test.cpp:53]: Remove commented code

File: src/main.cpp [hunk 6]
```diff
@@ -1,35 +1,57 @@
-#include <iostream>
-#include <vector>
+//#define CATCH_CONFIG_MAIN
+//#include <catch2/catch.hpp>
+#include <eigen3/Eigen/Dense>
 #include "../include/Raster.h"
```
Comment (c-bik) [inline src/main.cpp:4]: Unused include?

File: CMakeLists.txt [hunk 7]
```diff
@@ -7,10 +7,13 @@ set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -O0")
...
 else(GDAL_FOUND)
-  message(WARNING "GDAL NOT FOUND. Skipping GDAL package build...")
+    message(WARNING "GDAL NOT FOUND. Skipping GDAL package build...")
 endif()
 
+add_subdirectory(dependencies)
+add_subdirectory(tests)
+
 add_executable(openGCAS src/main.cpp src/Raster.cpp src/PolySelect.cpp)
 target_link_libraries(openGCAS PRIVATE GDAL::GDAL)
```
Comment (c-bik) [inline CMakeLists.txt:19]: New line at EOF

**Review by c-bik — CHANGES_REQUESTED** [review]
Docker breaks on this PR. Needs to fixed and ideally tests should be run in docker as well.

File: tests/raster_test.cpp (hunk 4 above)
Comment (quothbonney) [reply tests/raster_test.cpp:54]: Unsure of what you meant by this. Fixed the rest.

Comment (c-bik) [reply tests/raster_test.cpp:54]: I was hoping there is a way to specify data for testing externally. I need to read about CMake tests to see if it's possible. All good for now.

**Review by c-bik — APPROVED** [review]

**Pull request merged**